### PR TITLE
NetCDF: Build against HDF5 1.14

### PR DESCRIPTION
**Pull request by eschnett**
- update version_offset
- correct tarball checksum
- correct Windows build instruction
- remove outdated Apple work-around
- build on more platforms
- add new dependencies bzip2 (necessary) and zstd (because we can)

### CODE REVIEW

**Comment by giordano** [comment]
CC @Alexander-Barth

File: N/NetCDF/build_tarballs.jl [hunk 1]
```diff
@@ -20,31 +20,29 @@ version = VersionNumber(upstream_version.major * 100 + version_offset.major,
...
+cd $WORKSPACE/srcdir/netcdf-c*
 
 export CPPFLAGS="-I${includedir}"
 export LDFLAGS="-L${libdir}"
 export LDFLAGS_MAKE="${LDFLAGS}"
 CONFIGURE_OPTIONS=""
 
 if [[ ${target} == *-mingw* ]]; then
-    export LIBS="-lhdf5-0 -lhdf5_hl-0 -lcurl-4 -lz"
+    # we should determine the dll version (?) automatically
```
Comment (giordano) [inline N/NetCDF/build_tarballs.jl:37]: Yeah, to be honest I'm a bit worried about those random numbers like 310.  Where they come from?  Will they change in hdf5 1.15?  If so, it'd be ABI-incompatible with v1.14

Comment (eschnett) [reply N/NetCDF/build_tarballs.jl:37]: I think that's essentially the [soversion](https://cmake.org/cmake/help/latest/prop_tgt/SOVERSION.html) of the library, produced by HDF5's build system. We can probably, if we want to, modify the Windows build script for HDF5 to change these numbers to `0`.

On the other hand, if these numbers indicate binary (ABI) compatibility, then we should probably heed them, and be quite strict about the `compat` declaration when using `HDF5_jll`.

Comment (giordano) [reply N/NetCDF/build_tarballs.jl:37]: > On the other hand, if these numbers indicate binary (ABI) compatibility

Yes, the SONAME is part of the ABI, whenever it changes the ABI changes.  If they change the soname for every minor version (sigh) we really must be strict about compat bounds.

File: N/NetCDF/build_tarballs.jl [hunk 2]
```diff
@@ -98,13 +84,14 @@ products = [
 
 # Dependencies that must be installed before this package can be built
 dependencies = [
-    Dependency(PackageSpec(name="HDF5_jll"), compat="~1.12.2"),
-    Dependency("Zlib_jll"),
-    Dependency("XML2_jll"),
+    Dependency("Bzip2_jll"),
+    Dependency("HDF5_jll"; compat = "1.14"),
```
Comment (mkitti) [inline N/NetCDF/build_tarballs.jl:88]: We might want 
```suggestion
    Dependency("HDF5_jll"; compat = "~1.14"),
```

See
```
julia> Pkg.Types.semver_spec("1.14") |> print
1.14.0-1
julia> Pkg.Types.semver_spec("~1.14.0") |> print
1.14
```

**Pull request merged**